### ISSUE TITLE
Snapclient - no sound when running as service

### ISSUE THREAD

**Issue by feerlessleadr** [badaix/snapcast#521]
Hi - I've searched google and this repo and have tried a number of fixes suggested to no avail. 

I just purchased a pi zero w, along with a USB 3.5mm audio card to attach to some speakers for multi room audio. I also have a spare android phone that I use the snapdroid app on to test as another snapclient. 

When I run the snapclient in daemon mode on my pi, I get no sound coming out of the connected speaker via the usb card. However when I run snapclient manually, music plays out of my usb card to my speaker no problem. I have tried copying the settings from the manual run into /etc/default/snapclient, but that does not solve my issue. 

Please note that the music plays from my android phone, no problem. 

Here is my snapclient config at `/etc/default/snapclient`

`SNAPCLIENT_OPTS="--user snapclient:audio -h 192.168.1.17"`

Here is the manual inputs when I run snapclient in manual mode:

`snapclient --user snapclient:audio -h 192.168.1.17`

I'm not seeing any errors with the snapclient service:

```
pi@raspberrypi-1:~ $ sudo systemctl status snapclient
● snapclient.service - Snapcast client
   Loaded: loaded (/lib/systemd/system/snapclient.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2019-12-30 10:01:41 EST; 2s ago
     Docs: man:snapclient(1)
 Main PID: 1008 (snapclient)
   Memory: 1.6M
   CGroup: /system.slice/snapclient.service
           └─1008 /usr/bin/snapclient --user snapclient:audio -h 192.168.1.17

Dec 30 10:01:41 raspberrypi-1 systemd[1]: Started Snapcast client.
Dec 30 10:01:42 raspberrypi-1 snapclient[1008]: Connected to 192.168.1.17
```

Here is the output from running the snapclient manually and play a song for a few seconds and then stop:

```
pi@raspberrypi-1:~ $ snapclient --user snapclient:audio -h 192.168.1.17
2019-12-30 10-04-47 [Info] Latency: 0
2019-12-30 10-04-47 [Notice] Connected to 192.168.1.17
2019-12-30 10-04-47 [Info] My MAC: "b8:27:eb:76:3e:eb", socket: 6
2019-12-30 10-04-48 [Info] ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
metadata:{"STREAM":"default"}
2019-12-30 10-04-48 [Info] Codec: flac
2019-12-30 10-04-48 [Notice] sampleformat: 48000:16:2
2019-12-30 10-04-48 [Info] frames: 1440
2019-12-30 10-04-48 [Info] age > 200: -947
2019-12-30 10-04-48 [Info] sleep < -bufferDuration/2: -947 < -15, sleep: -938
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -938 < -15, sleep: -916
2019-12-30 10-04-49 [Err] XRUN
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -916 < -15, sleep: -889
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -889 < -15, sleep: -866
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -866 < -15, sleep: -835
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -835 < -15, sleep: -798
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -798 < -15, sleep: -750
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -750 < -15, sleep: -714
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -714 < -15, sleep: -668
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -668 < -15, sleep: -633
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -633 < -15, sleep: -584
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -584 < -15, sleep: -525
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -525 < -15, sleep: -476
2019-12-30 10-04-49 [Info] diff to server [ms]: 1.617
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -476 < -15, sleep: -438
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -438 < -15, sleep: -401
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -401 < -15, sleep: -365
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -365 < -15, sleep: -331
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -331 < -15, sleep: -298
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -298 < -15, sleep: -266
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -266 < -15, sleep: -236
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -236 < -15, sleep: -206
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -206 < -15, sleep: -177
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -177 < -15, sleep: -150
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -150 < -15, sleep: -123
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -123 < -15, sleep: -96
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -96 < -15, sleep: -69
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -69 < -15, sleep: -43
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -43 < -15, sleep: -20
2019-12-30 10-04-49 [Info] sleep < -bufferDuration/2: -20 < -15, sleep: 5
2019-12-30 10-04-49 [Info] Sleep 5, age: 8, bufferDuration: 30
2019-12-30 10-04-49 [Info] Chunk: 80    80      80      80      1       249
2019-12-30 10-04-49 [Info] Sleep 5, age: 0, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -7, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -12, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -19, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -24, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -28, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -35, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 5, age: -40, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 4, age: -44, bufferDuration: 30
2019-12-30 10-04-49 [Info] Sleep 4, age: -50, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -53, bufferDuration: 30
2019-12-30 10-04-50 [Info] Chunk: -539  -539    -539    -539    1       186
2019-12-30 10-04-50 [Info] Sleep 4, age: -57, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -61, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -66, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -70, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -73, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -76, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 4, age: -80, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -83, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -86, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -88, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -91, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -94, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -96, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -98, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -100, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -101, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 3, age: -103, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 2, age: -104, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 2, age: -105, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 2, age: -106, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 2, age: -107, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 2, age: -108, bufferDuration: 30
2019-12-30 10-04-50 [Info] Sleep 1, age: -107, bufferDuration: 30
2019-12-30 10-04-51 [Info] Chunk: -1071 -1071   -1071   -1071   1       133
2019-12-30 10-04-51 [Info] Sleep 1, age: -106, bufferDuration: 30
2019-12-30 10-04-51 [Info] Sleep 0, age: -105, bufferDuration: 30
2019-12-30 10-04-51 [Info] Sleep 0, age: -106, bufferDuration: 30
2019-12-30 10-04-51 [Info] Sleep 0
2019-12-30 10-04-52 [Info] Chunk: -1067 -1075   -1075   -1075   20      149
2019-12-30 10-04-52 [Info] pMiniBuffer->full() && (abs(pMiniBuffer->mean()) > 50): -107532
2019-12-30 10-04-52 [Info] sleep < -bufferDuration/2: -107 < -15, sleep: -105
2019-12-30 10-04-52 [Info] sleep < -bufferDuration/2: -105 < -15, sleep: -75
2019-12-30 10-04-52 [Info] sleep < -bufferDuration/2: -75 < -15, sleep: -43
2019-12-30 10-04-52 [Info] sleep < -bufferDuration/2: -43 < -15, sleep: -14
2019-12-30 10-04-52 [Info] Sleep -14, age: -10, bufferDuration: 30
2019-12-30 10-04-52 [Info] Sleep -14, age: -14, bufferDuration: 30
2019-12-30 10-04-52 [Info] Sleep -14, age: -13, bufferDuration: 30
2019-12-30 10-04-52 [Info] Sleep -14, age: -12, bufferDuration: 30
2019-12-30 10-04-52 [Info] Sleep -13, age: -11, bufferDuration: 30
2019-12-30 10-04-52 [Info] Sleep -13, age: -10, bufferDuration: 30
2019-12-30 10-04-53 [Info] Chunk: -105  -105    -105    -105    1       147
2019-12-30 10-04-53 [Info] Sleep -10, age: -9, bufferDuration: 30
2019-12-30 10-04-53 [Info] Sleep -10, age: -8, bufferDuration: 30
2019-12-30 10-04-53 [Info] Sleep -9, age: -7, bufferDuration: 30
2019-12-30 10-04-53 [Info] Sleep -9, age: -6, bufferDuration: 30
2019-12-30 10-04-53 [Info] Sleep -8, age: -5, bufferDuration: 30
2019-12-30 10-04-54 [Info] Chunk: -54   -54     -54     -54     1       141
2019-12-30 10-04-54 [Info] Sleep -7, age: -4, bufferDuration: 30
2019-12-30 10-04-54 [Info] Sleep -7, age: -2, bufferDuration: 30
2019-12-30 10-04-54 [Info] Sleep -7, age: -4, bufferDuration: 30
2019-12-30 10-04-54 [Info] Sleep -5, age: -3, bufferDuration: 30
2019-12-30 10-04-55 [Info] Chunk: -30   -30     -30     -30     1       147
2019-12-30 10-04-55 [Info] Sleep -5, age: -2, bufferDuration: 30
2019-12-30 10-04-55 [Info] Sleep -4, age: -1, bufferDuration: 30
2019-12-30 10-04-55 [Info] Sleep -4, age: 0, bufferDuration: 30
2019-12-30 10-04-55 [Info] Sleep -2, age: 1, bufferDuration: 30
2019-12-30 10-04-56 [Info] Chunk: 18    18      18      18      1       146
2019-12-30 10-04-56 [Info] Sleep -1, age: 2, bufferDuration: 30
2019-12-30 10-04-56 [Info] Sleep 0, age: 3, bufferDuration: 30
2019-12-30 10-04-56 [Info] Sleep 0
2019-12-30 10-04-57 [Info] Chunk: 34    35      35      35      15      145
2019-12-30 10-04-58 [Info] Chunk: 31    34      34      34      48      145
2019-12-30 10-04-59 [Info] Chunk: 24    25      34      34      81      143
2019-12-30 10-05-00 [Info] Chunk: 39    34      30      32      114     146
2019-12-30 10-05-01 [Info] Chunk: 31    33      30      33      148     143
2019-12-30 10-05-01 [Info] Failed to get chunk
2019-12-30 10-05-06 [Notice] No chunk received for 5000ms. Closing ALSA.
```

I'm not a linux expert by any means, so hopefully someone has an idea what I'm doing wrong. Thanks

**Comment by lowb1rd** [comment]
Make sure that this line is present in your snapclient config:
`START_SNAPCLIENT=true`

**Comment by feerlessleadr** [comment]
Thanks - just added that into the client config, so it now looks like so:

```
START_SNAPCLIENT=true
SNAPCLIENT_OPTS="--user snapclient:audio -h 192.168.1.17"
```

No change unfortunately though. playing through the daemon has no sound, whereas manually running it works fine

**Comment by fuchsmich** [comment]
Could you also post your `/lib/systemd/system/snapclient.service`?

I don't think you need to specify `--user` if using the systemd service for starting snapclient. The user and group, which run the process, are defined in the [snapclient.service](https://github.com/badaix/snapcast/blob/865701aa0fc7cea209e009cb6d1173a1f2f59ed4/debian/snapclient.service#L10) file. As you can see the group is `snapclient` not `audio`.

So to emulate the startup from CLI you could try:

> sudo -u snapclient -g snapclient snapclient -h 192.168.1.17

If there is still no sound you could change the group in your `snapclient.service` to `audio`. Reload the changes in the file `systemctl daemon-reload` and restart the service `systemctl restart snapclient` (or restart your rpi).

**Comment by feerlessleadr** [comment]
Thanks - here is what the snapclient service looks like

```
[Unit]
Description=Snapcast client
Documentation=man:snapclient(1)
Wants=avahi-daemon.service
After=network.target time-sync.target sound.target avahi-daemon.service

[Service]
EnvironmentFile=-/etc/default/snapclient
ExecStart=/usr/bin/snapclient $SNAPCLIENT_OPTS
User=snapclient
Group=snapclient
# very noisy on stdout
StandardOutput=null
Restart=on-failure

[Install]
WantedBy=multi-user.target
```
I don't know what I did when I was fiddling with this over the last few days, but now even when I run snapclient manually, I get no sound from my pi zero. 

I was trying a bunch of different things from multiple guides, and I think I messed something up even further. I'm going to try a fresh image of raspbian.

**Comment by fuchsmich** [comment]
One simple test would be running.

    $ sudo -u snapclient -g snapclient snapclient -l

This should list some alsa devices.
If not, check if user `snapclient` is in groups `audio`.

    $ groups snapclient

Output should be:

     audio snapclient

**Comment by feerlessleadr** [comment]
Ok - I re-installed raspbian on my zero w, re-followed [this guide](https://github.com/skalavala/Multi-Room-Audio-Centralized-Audio-for-Home) and [this guide](https://medium.com/@robert.guiscard/raspberry-pi-zero-wh-with-external-usb-sound-card-28e66d7d103e) (to set my USB soundcard as default on the system) and am back to being able to play music via manual run of snapclient and not via daemon when running manually with the below command (with the daemon stopped):

`snapclient --user snapclient:audio -h 192.168.1.17`

I tried running the command you suggested, but I get no sound.

`sudo -u snapclient -g snapclient snapclient -h 192.168.1.17`

However, when I run the following command, the sound works:

`snapclient -u snapclient -g snapclient -h 192.168.1.17`

As far as I can tell, the only difference is running as sudo vs no sudo. Not sure if that sheds any light as to why the daemon doesn't work. 

I then tried updating the `/lib/systemd/system/snapclient.service` with the new audio group, reloaded the daemon, restarted the service, but still no sound. Here is the updated `snapclient.service`:

```
[Unit]
Description=Snapcast client
Documentation=man:snapclient(1)
Wants=avahi-daemon.service
After=network.target time-sync.target sound.target avahi-daemon.service

[Service]
EnvironmentFile=-/etc/default/snapclient
ExecStart=/usr/bin/snapclient $SNAPCLIENT_OPTS
User=snapclient
Group=audio
# very noisy on stdout
StandardOutput=null
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

Could it possibly have anything to do with my usb soundcard not being recognized by the daemon?

**Comment by feerlessleadr** [comment]
@fuchsmich Apologies, I posted my reply without realizing you had responded. 

Here is the output of `sudo -u snapclient -g snapclient snapclient -l`
```
pi@raspberrypi-1:~ $ sudo -u snapclient -g snapclient snapclient -l
0: null
Discard all samples (playback) or generate zero samples (capture)

1: jack
JACK Audio Connection Kit

2: pulse
PulseAudio Sound Server

3: default:CARD=ALSA
bcm2835 ALSA, bcm2835 ALSA
Default Audio Device

4: sysdefault:CARD=ALSA
bcm2835 ALSA, bcm2835 ALSA
Default Audio Device

5: dmix:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct sample mixing device

6: dmix:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct sample mixing device

7: dmix:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct sample mixing device

8: dsnoop:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct sample snooping device

9: dsnoop:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct sample snooping device

10: dsnoop:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct sample snooping device

11: hw:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct hardware device without any conversions

12: hw:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct hardware device without any conversions

13: hw:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct hardware device without any conversions

14: plughw:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Hardware device with all software conversions

15: plughw:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Hardware device with all software conversions

16: plughw:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Hardware device with all software conversions

17: usbstream:CARD=ALSA
bcm2835 ALSA
USB Stream Output

18: default:CARD=Device
USB Audio Device, USB Audio
Default Audio Device

19: sysdefault:CARD=Device
USB Audio Device, USB Audio
Default Audio Device

20: front:CARD=Device,DEV=0
USB Audio Device, USB Audio
Front speakers

21: surround21:CARD=Device,DEV=0
USB Audio Device, USB Audio
2.1 Surround output to Front and Subwoofer speakers

22: surround40:CARD=Device,DEV=0
USB Audio Device, USB Audio
4.0 Surround output to Front and Rear speakers

23: surround41:CARD=Device,DEV=0
USB Audio Device, USB Audio
4.1 Surround output to Front, Rear and Subwoofer speakers

24: surround50:CARD=Device,DEV=0
USB Audio Device, USB Audio
5.0 Surround output to Front, Center and Rear speakers

25: surround51:CARD=Device,DEV=0
USB Audio Device, USB Audio
5.1 Surround output to Front, Center, Rear and Subwoofer speakers

26: surround71:CARD=Device,DEV=0
USB Audio Device, USB Audio
7.1 Surround output to Front, Center, Side, Rear and Woofer speakers

27: iec958:CARD=Device,DEV=0
USB Audio Device, USB Audio
IEC958 (S/PDIF) Digital Audio Output

28: dmix:CARD=Device,DEV=0
USB Audio Device, USB Audio
Direct sample mixing device

29: dsnoop:CARD=Device,DEV=0
USB Audio Device, USB Audio
Direct sample snooping device

30: hw:CARD=Device,DEV=0
USB Audio Device, USB Audio
Direct hardware device without any conversions

31: plughw:CARD=Device,DEV=0
USB Audio Device, USB Audio
Hardware device with all software conversions

32: usbstream:CARD=Device
USB Audio Device
USB Stream Output
```
It also looks like the groups are correct:
```
pi@raspberrypi-1:~ $ groups snapclient
snapclient : snapclient audio
```

**Comment by fuchsmich** [comment]
> You need to set up Pulse Audio in every Raspberry Pi that plays music to the connected speaker.

This is definitly false. No need for pulse if your devices only job is playing sound from snapclient. Try again without pulse.

**Comment by feerlessleadr** [comment]
ok - so I got it working! thank you @fuchsmich 

Here is what I did that ultimately got it to work (also to remind me what I did when I need to set this up again). I created a new image on a new SD card to test. I only installed snapclient, and did not touch anything to do with pulse. However, I still wasn't getting any sound (either via daemon or manually running it). 

I re-ran the following command to list the pcm devices

`sudo -u snapclient -g snapclient snapclient -l`

Then I grabbed the index number for my usb card (trial and error), which turned out to be the following:
```
15: default:CARD=Device
USB Audio Device, USB Audio
Default Audio Device
```
when I added in the -s 15 option to the snapclient config, the daemon now plays music as expected.

Thanks again for your help!
